File: backend/src/main/java/me/performancereservation/domain/performance/entities/PerformanceSchedule.java
package me.performancereservation.domain.performance.entities;

import jakarta.persistence.*;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import me.performancereservation.domain.common.BaseEntity;
import me.performancereservation.global.exception.ErrorCode;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor
public class PerformanceSchedule extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Long performanceId; // (FK) 공연 ID

    private LocalDateTime startTime; // 공연 시작 시간

    private LocalDateTime endTime; // 공연 종료 시간

    private int remainingSeats; // 남은 좌석

    @Column(name = "is_canceled")
    private boolean canceled; // 회차 취소 여부

    @Builder
    public PerformanceSchedule(Long id, Long performanceId, LocalDateTime startTime, LocalDateTime endTime, int remainingSeats, boolean canceled) {
        this.id = id;
        this.performanceId = performanceId;
        this.startTime = startTime;
        this.endTime = endTime;
        this.remainingSeats = remainingSeats;
        this.canceled = canceled;
    }

    public void cancel() {
        this.canceled = true;
    }

    public void decreaseRemainingSeats(int quantity) {
        if(this.remainingSeats < quantity) {
            throw ErrorCode.NO_REMAINING_SEATS.domainException("잔여 좌석이 부족합니다. remainingSeats = " + this.remainingSeats);
        }
        this.remainingSeats -= quantity;
    }

    public boolean hasPermission(Long performanceId) {
        return performanceId != null && performanceId.equals(this.performanceId);
    }
}

File: backend/src/main/kotlin/me/performancereservation/domain/performance/entities/PerformanceSchedule.kt
package me.performancereservation.domain.performance.entities

import jakarta.persistence.*
import me.performancereservation.domain.common.BaseEntity
import me.performancereservation.global.exception.ErrorCode
import java.time.LocalDateTime

@Entity
class PerformanceSchedule (
    // (PK) 회차 ID
    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,
    // (FK) 공연 ID
    var performanceId: Long,
    // 공연 시작 시간
    var startTime: LocalDateTime,
    // 공연 종료 시간
    var endTime: LocalDateTime,
    // 남은 좌석
    var remainingSeats: Int,
    // 회차 취소 여부
    @field:Column(name = "is_canceled")
    var canceled: Boolean
) : BaseEntity() {

    fun cancel() {
        this.canceled = true
    }

    fun decreaseRemainingSeats(quantity: Int) {
        if (this.remainingSeats < quantity) {
            throw ErrorCode.NO_REMAINING_SEATS.domainException("잔여 좌석이 부족합니다. remainingSeats = " + this.remainingSeats)
        }
        this.remainingSeats -= quantity
    }

    fun increaseRemainingSeats(quantity: Int) {
        if (quantity <= 0) {
            throw ErrorCode.NO_REMAINING_SEATS.domainException("좌석수가 1이상이 아님")
        }

        this.remainingSeats += quantity
    }

    fun hasPermission(performanceId: Long?): Boolean {
        return performanceId != null && performanceId == this.performanceId
    }

    override fun equals(other: Any?): Boolean {
        if(this === other) return true
        if(other !is PerformanceSchedule) return false

        return id == other.id && updatedAt == other.updatedAt
    }

    override fun hashCode(): Int {
        var result = id.hashCode()

        result = 31 * result + updatedAt.hashCode()

        return result
    }
}